File: tigerline/stake.py
"""Stake allocator — A+/A/B/C/D → bankroll percentage → stake amount.

The level is picked by the recommender from (classifier confidence + scenario).
The harness adjustment then nudges the % within the level's band: ``upgrade``
clamps to max_pct, ``downgrade`` clamps to min_pct, ``normal`` uses default_pct,
``skip`` zeroes everything regardless of level.
"""

from __future__ import annotations

from decimal import ROUND_HALF_UP, Decimal

from tigerline.models import Adjustment, StakeLevel
from tigerline.rules import StakeBand, load_stake_bands


def allocate(
    level: StakeLevel,
    bankroll: Decimal,
    adjustment: Adjustment = "normal",
    bands: dict[StakeLevel, StakeBand] | None = None,
) -> tuple[Decimal, Decimal]:
    """Return ``(stake_pct, stake_amount)`` for ``level`` under ``adjustment``.

    The amount is rounded to the nearest whole currency unit (Decimal "1").
    Skipped bets always come back as (0, 0).
    """
    if bands is None:
        bands = load_stake_bands()

    band = bands[level]
    if adjustment == "skip" or level == "D":
        return Decimal("0"), Decimal("0")

    if adjustment == "upgrade":
        pct = band.max_pct
    elif adjustment == "downgrade":
        pct = band.min_pct
    else:
        pct = band.default_pct

    amount = (bankroll * pct).quantize(Decimal("1"), rounding=ROUND_HALF_UP)
    return pct, amount


def pick_level(scenario_confidence: Decimal, adjustment: Adjustment) -> StakeLevel:
    """Map (confidence, adjustment) → stake level for the main bet.

    The recommender uses this when the classifier produces a real scenario.
    Confidence bands roughly match the rule_confidence heuristic in classifier.py.
    """
    if adjustment == "skip":
        return "D"

    if scenario_confidence >= Decimal("0.85"):
        base: StakeLevel = "A+" if adjustment == "upgrade" else "A"
    elif scenario_confidence >= Decimal("0.55"):
        base = "A" if adjustment == "upgrade" else "B"
    elif scenario_confidence >= Decimal("0.35"):
        base = "B" if adjustment == "upgrade" else "C"
    else:
        base = "C" if adjustment == "upgrade" else "D"

    return base


__all__ = ["allocate", "pick_level"]

File: tigerline-web/js/stake.js
/* =====================================================================
 * stake.js — level → bankroll %. Ports tigerline/stake.py +
 * config/stake_levels.yaml.
 * ===================================================================== */
"use strict";

import { dec, roundHalfUp } from "./decimal.js";

export const STAKE_BANDS = {
  "A+": { min_pct: 0.18, max_pct: 0.25, default_pct: 0.20 },
  A:    { min_pct: 0.12, max_pct: 0.18, default_pct: 0.15 },
  B:    { min_pct: 0.07, max_pct: 0.12, default_pct: 0.09 },
  C:    { min_pct: 0.03, max_pct: 0.06, default_pct: 0.04 },
  D:    { min_pct: 0,    max_pct: 0,    default_pct: 0    },
};

/**
 * @param {"A+"|"A"|"B"|"C"|"D"} level
 * @param {number} bankroll
 * @param {"upgrade"|"normal"|"downgrade"|"skip"} adjustment
 * @returns {{pct:number, amount:number}}
 */
export function allocate(level, bankroll, adjustment = "normal") {
  if (adjustment === "skip" || level === "D") return { pct: 0, amount: 0 };
  const band = STAKE_BANDS[level];
  const pct = adjustment === "upgrade" ? band.max_pct
           : adjustment === "downgrade" ? band.min_pct
           : band.default_pct;
  return { pct, amount: roundHalfUp(bankroll * pct) };
}

/**
 * @param {number} confidence
 * @param {"upgrade"|"normal"|"downgrade"|"skip"} adjustment
 * @returns {"A+"|"A"|"B"|"C"|"D"}
 */
export function pickLevel(confidence, adjustment) {
  if (adjustment === "skip") return "D";
  if (dec.gte(confidence, 0.85)) return adjustment === "upgrade" ? "A+" : "A";
  if (dec.gte(confidence, 0.55)) return adjustment === "upgrade" ? "A"  : "B";
  if (dec.gte(confidence, 0.35)) return adjustment === "upgrade" ? "B"  : "C";
  return adjustment === "upgrade" ? "C" : "D";
}
